ReactAndroid/src/main/java/com/facebook/react/bridge/BaseJavaModule.java

```diff
@@ -49,17 +54,35 @@
  */
 public abstract class BaseJavaModule implements NativeModule {
   // taken from Libraries/Utilities/MessageQueue.js
-  static final public String METHOD_TYPE_ASYNC = "async";
-  static final public String METHOD_TYPE_PROMISE= "promise";
+  private static final String METHOD_TYPE_ASYNC = "async";
+  private static final String METHOD_TYPE_PROMISE= "promise";
   static final public String METHOD_TYPE_SYNC = "sync";
+  private final List<? extends ArgumentExtractor.Factory> mArgumentExtractorFactories;
+
+  public BaseJavaModule() {
+    this(Collections.<ArgumentExtractor.Factory>emptyList());
+  }
 
-  private static abstract class ArgumentExtractor<T> {
+  public BaseJavaModule(List<? extends ArgumentExtractor.Factory> argumentExtractorFactories) {
+    mArgumentExtractorFactories = argumentExtractorFactories;
+  }
+
+  public static abstract class ArgumentExtractor<T> {
     public int getJSArgumentsNeeded() {
       return 1;
     }
 
     public abstract @Nullable T extractArgument(
         CatalystInstance catalystInstance, ExecutorToken executorToken, ReadableNativeArray jsArguments, int atIndex);
+
+    /**
+     * Returns an instance of {@link ArgumentExtractor} that is able to handle extracting arguments
+     * to {@link ReactMethod} annotated methods with the given {@code Type}. Return {@code null} if
+     * this factory can't handle the provided type.
+     */
+    public interface Factory {
```

Comment:
This allows you to "plug-in" the logic that associates a specific `ArgumentExtractor` with a `Type`. This way you wouldn't need the big `if/else/else if...` chain in `buildArgumentExtractors()` and could delegate it to the Factories to provide or not an extractor for the provided Type.